### PR TITLE
Add CloudEvent binding

### LINKED ISSUE #20580: Support CloudEvent in functions extensions
Since the new SDK supports this, we should add bindings for the extensions as well.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/20580

### CODE REVIEW

File: sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/EventGridExtensionConfigProvider.cs [hunk 1]
```diff
@@ -85,8 +88,30 @@ public void Initialize(ExtensionConfigContext context)
...
                 .AddBindingRule<EventGridAttribute>()
                 //TODO - add binding for BinaryData?
-                .AddConverter<string, EventGridEvent>((str) => EventGridEvent.Parse(new BinaryData(str)))
-                .AddConverter<JObject, EventGridEvent>((jobject) =>  EventGridEvent.Parse(new BinaryData(jobject.ToString())));
+                .AddConverter<string, object>(str =>
+                {
+                    // first attempt to parse as EventGridEvent, then fallback to CloudEvent
+                    try
+                    {
+                        return EventGridEvent.Parse(new BinaryData(str));
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/EventGridExtensionConfigProvider.cs:96]: Any way we can try to parse it without throwing an exception?

Comment (JoshLove-msft) [reply sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/EventGridExtensionConfigProvider.cs:96]: We could, but I'd rather not recreate the Parse logic here. We can potentially add TryParse methods to the EG SDK.

Comment (pakrym) [reply sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/EventGridExtensionConfigProvider.cs:96]: Let's file an issue

Comment (JoshLove-msft) [reply sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/src/EventGridExtensionConfigProvider.cs:96]: https://github.com/Azure/azure-sdk-for-net/issues/20702

File: sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/tests/JobhostEndToEnd.cs [hunk 2]
```diff
@@ -277,6 +333,122 @@ public void TestEventGridToCustomCollection_Batch([EventGridTrigger] ValidPoco[]
...
+        [TestCase("StringEvents", "0 1 2 3 4")]
+        [TestCase("JObjectEvents", "0 1 2 3 4")]
+        public async Task OutputCloudEventBindingParamsTests(string functionName, string expectedCollection)
+        {
+            List<CloudEvent> cloudEvents = new List<CloudEvent>();
+            List<EventGridEvent> egEvents = new List<EventGridEvent>();
+
+            Func<EventGridAttribute, IAsyncCollector<object>> eventConverter = (attr =>
+            {
+                var mockClient = new Mock<EventGridPublisherClient>();
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/tests/JobhostEndToEnd.cs:354]: This test seems to test the cloud event only. Why do we need to mock both methods?

**Review by pakrym — APPROVED** [review]
Sweet!

File: sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/tests/JobhostEndToEnd.cs (hunk 2 above)
Comment (pakrym) [reply sdk/eventgrid/Microsoft.Azure.WebJobs.Extensions.EventGrid/tests/JobhostEndToEnd.cs:354]: There is another test like this one.

**Pull request merged**